src/node_crypto.cc

```diff
@@ -2210,49 +2242,91 @@ void Connection::Initialize(Environment* env, Handle<Object> target) {
-  //   By applying X509_CTX_store_* functions VerifyCallback can locate the
-  //   certificate in question and perform additional steps (see EXAMPLES).
-  //   If no error is found for a certificate, VerifyCallback is called
-  //   with preverify_ok=1 before advancing to the next level.
-  //
-  //   The return value of VerifyCallback controls the strategy of the
-  //   further verification process. If VerifyCallback returns 0, the
-  //   verification process is immediately stopped with "verification
-  //   failed" state. If SSL_VERIFY_PEER is set, a verification failure
-  //   alert is sent to the peer and the TLS/SSL handshake is terminated. If
-  //   VerifyCallback returns 1, the verification process is continued. If
-  //   VerifyCallback always returns 1, the TLS/SSL handshake will not be
-  //   terminated with respect to verification failures and the connection
-  //   will be established. The calling process can however retrieve the
-  //   error code of the last verification error using
-  //   SSL_get_verify_result(3) or by maintaining its own error storage
-  //   managed by VerifyCallback.
-  //
-  //   If no VerifyCallback is specified, the default callback will be
-  //   used.  Its return value is identical to preverify_ok, so that any
-  //   verification failure will lead to a termination of the TLS/SSL
-  //   handshake with an alert message, if SSL_VERIFY_PEER is set.
-  //
-  // Since we cannot perform I/O quickly enough in this callback, we ignore
-  // all preverify_ok errors and let the handshake continue. It is
-  // imparative that the user use Connection::VerifyError after the
-  // 'secure' callback has been made.
-  return 1;
+inline int compar(const void* a, const void* b) {
+  return memcmp(a, b, CNNIC_WHITELIST_HASH_LEN);
+}
+
+
+inline int IsSelfSigned(X509* cert) {
+  return X509_NAME_cmp(X509_get_subject_name(cert),
+                       X509_get_issuer_name(cert)) == 0;
+}
+
+
+inline X509* FindRoot(STACK_OF(X509)* sk) {
+  for (int i = 0; i < sk_X509_num(sk); i++) {
+    X509* cert = sk_X509_value(sk, i);
+    if (IsSelfSigned(cert))
+      return cert;
+  }
+  return nullptr;
+}
+
+
+// Whitelist check for certs issued by CNNIC. See
+// https://blog.mozilla.org/security/2015/04/02
+// /distrusting-new-cnnic-certificates/
+inline CheckResult CheckWhitelistedServerCert(X509_STORE_CTX* ctx) {
+  unsigned char hash[CNNIC_WHITELIST_HASH_LEN];
+  unsigned int hashlen = CNNIC_WHITELIST_HASH_LEN;
+
+  STACK_OF(X509)* chain = X509_STORE_CTX_get1_chain(ctx);
+  CHECK_NE(chain, nullptr);
+  CHECK_GT(sk_X509_num(chain), 0);
+
+  // Take the last cert as root at the first time.
+  X509* root_cert = sk_X509_value(chain, sk_X509_num(chain)-1);
+  X509_NAME* root_name = X509_get_subject_name(root_cert);
+
+  if (!IsSelfSigned(root_cert)) {
+    root_cert = FindRoot(chain);
+    CHECK_NE(root_cert, nullptr);
+    root_name = X509_get_subject_name(root_cert);
+  }
```

Comment:
@shigeki Do you remember why you wrote it like this? Unless I misunderstand something a valid chain contains only one self-signed certificate and it's always the last one, isn't it?

Reply:
@bnoordhuis  Honestly I did not remember it. But I guess I wrote it in order to care about the alt cert chain which has a cross cert in its chain as below. When the SHA-1 root was deleted in the root store, I was just afraid of the case when the last one in the cert chain stack was not SHA-2 root but its cross root to SHA-1.
![alt_chain 001](https://user-images.githubusercontent.com/782880/54040625-2d14fd80-4209-11e9-9809-04666f07350f.jpeg)